### PR TITLE
Implemented tags and taggers

### PR DESCRIPTION
**Pull request by surprisetalk**
Note: This PR only implements `#a () #b ()` and not its corresponding "sugar" form `#a #b`.

### CODE REVIEW

**Comment by tekknolagi** [comment]
I have held off on looking at this because you marked it as a draft but if you would like feedback, please let me know!

**Comment by surprisetalk** [comment]
@tekknolagi Yes, I'd love a vibe check if you could spare the time :)

File: scrapscript.py [hunk 1]
```diff
@@ -473,9 +478,17 @@ def parse(tokens: typing.List[Token], p: float = 0) -> "Object":
             break
         tokens.pop(0)
         if op == Operator("="):
-            if not isinstance(l, Var):
+            if isinstance(l, Binop) and BinopKind.HASTYPE == l.op:
+                l = l.left  # TODO
```
Comment (tekknolagi) [inline scrapscript.py:482]: What's going on here? What is there to do? (please leave a note in the code)

File: scrapscript.py [hunk 2]
```diff
@@ -359,6 +349,8 @@ def parse_assign(tokens: typing.List[Token], p: float = 0) -> "Assign":
     assign = parse(tokens, p)
     if isinstance(assign, Spread):
         return Assign(Var("..."), assign)
+    if isinstance(assign, Binop) and assign.op == BinopKind.HASTYPE:
+        return Assign(Var("..."), assign)  # TODO
```
Comment (tekknolagi) [inline scrapscript.py:353]: What is there to do?

File: scrapscript.py [hunk 3]
```diff
@@ -653,6 +674,68 @@ def __str__(self) -> str:
...
+    @staticmethod
+    def deserialize(msg: Dict[str, object]) -> "Tags":
+        raise Exception("TODO")
+
+    def __str__(self) -> str:
+        return " ".join([f"#{k} {v}" for k, v in self.value.items()])
+
+
+@dataclass(eq=True, frozen=True, unsafe_hash=True)
+class Tagger(Object):
```
Comment (tekknolagi) [inline scrapscript.py:693]: What is a tagger? I think I understand what a `Tags` is, but these two things have similar names and no explanation (yet)

File: scrapscript.py [hunk 4]
```diff
@@ -1037,7 +1107,7 @@ def eval_list(env: Env, exp: Object) -> typing.List[Object]:
 
 
 def make_bool(x: bool) -> Object:
-    return Symbol("true" if x else "false")
+    return Tag(Tags({"true": Hole(), "false": Hole()}), "true" if x else "false", Hole())
```
Comment (tekknolagi) [inline scrapscript.py:1110]: this should probably be a global constant at this point and we can use in `eval_bool` as well

File: scrapscript.py [hunk 5]
```diff
@@ -1219,14 +1305,16 @@ def eval_exp(env: Env, exp: Object) -> Object:
...
+            isinstance(exp.binding, Binop) and exp.binding.op == BinopKind.HASTYPE
+        )
         res_env = eval_exp(env, exp.binding)
         assert isinstance(res_env, EnvObject)
         new_env = {**env, **res_env.env}
         return eval_exp(new_env, exp.body)
     if isinstance(exp, Assert):
         cond = eval_exp(env, exp.cond)
-        if cond != Symbol("true"):
+        if cond != Tag(Tags({"true": Hole(), "false": Hole()}), "true", Hole()):
```
Comment (tekknolagi) [inline scrapscript.py:1317]: as above with the global

**Review by tekknolagi — COMMENTED** [review]
The changes look sensible, I think, but the bigger picture is missing. Please document what's going on because I had to try and work it out from scratch while in your code. Especially this new generics thing, which I did not expect to see in a PR about tagged unions.

Also, I think Tags/Tagger/Tag is a little confusing name-wise. While that would be helped a little bit by documentation, I think changing the names to be more different is also a good idea

Re: tests: glad you added tests. We would like test names to be in the form `test_<action>_with/in_<situation>_returns/raises_<thing>` (though yes, we don't 100% have that in existing tests). And I think some of your test names say one thing but test another, which was confusing to me.

Last, nitpicking: when you leave TODOs, please explain what there is to do. It's impossible to do a drive-by cleanup when there's no context/breadcrumbs.

**Comment by surprisetalk** [comment]
This PR is stale! Closing it

**Pull request closed without merging**